src/reconcile/rec_write.c

```diff
@@ -1982,14 +1982,18 @@ __rec_split(WT_SESSION_IMPL *session, WT_RECONCILE *r, size_t next_len)
 		next->start = r->first_free;
 		next->entries = 0;
 
-		/*
-		 * Set the space available to another split-size chunk, if we
-		 * have one.  If we don't have room for another split chunk,
-		 * add whatever space remains in this page.
-		 */
+		/* Set the space available to another split-size chunk. */
 		r->space_avail =
 		    r->split_size - WT_PAGE_HEADER_BYTE_SIZE(btree);
-		if (inuse + r->space_avail > r->page_size) {
+
+		/*
+		 * If there is room but not enough room for a split size chunk
+		 * try to fit the rest of the content onto the page.
+		 * We may have already gathered more than a page worth of
+		 * data if there has been a large value.
+		 */
+		if (inuse <= r->page_size &&
+		    inuse + r->space_avail > r->page_size) {
 			WT_ASSERT(session, r->page_size >= inuse);
```

Comment:
Thanks, @agorrod, nice catch.

My concern about this change is we don't reset `r->space_avail`, it's still set to the next split size, so if `inuse` is greater than page size, but `next_len` is less than the split size, we'll keep on adding to the chunk.

What would you think of this change, instead?

The idea is that if we've grown bigger than a page, there's no space available, and we'll fall through into the `SPLIT_MAX` case immediately.

```
diff --git a/src/reconcile/rec_write.c b/src/reconcile/rec_write.c
index 4ca1a6b..cceaf71 100644
--- a/src/reconcile/rec_write.c
+++ b/src/reconcile/rec_write.c
@@ -1985,13 +1985,15 @@ __rec_split(WT_SESSION_IMPL *session, WT_RECONCILE *r, size_t next_len)
                /*
                 * Set the space available to another split-size chunk, if we
                 * have one.  If we don't have room for another split chunk,
-                * add whatever space remains in this page.
+                * add whatever space remains in this page, but be careful, in
+                * the case of a large item that forced us to grow the buffer
+                * early, we may have already crossed the page-size boundary.
                 */
                r->space_avail =
                    r->split_size - WT_PAGE_HEADER_BYTE_SIZE(btree);
                if (inuse + r->space_avail > r->page_size) {
-                       WT_ASSERT(session, r->page_size >= inuse);
-                       r->space_avail = r->page_size - inuse;
+                       r->space_avail =
+                           r->page_size > inuse ? (r->page_size - inuse) : 0;
 
                        /* There are no further boundary points. */
                        r->bnd_state = SPLIT_MAX;
```